changes to system test infrastructure and added tests to check output a…

dgergel opened the pull request:
…nd domain files for image driver

jhamman commented on tests/run_tests.py:
```diff
@@ -9,13 +9,17 @@
 from collections import OrderedDict
 import string
+import numpy as np
+import xarray as xr
```
can you add `xarray` and `bottleneck` to `ci/requirements.yml`?

jhamman commented on tests/run_tests.py:
```diff
@@ -292,7 +296,7 @@ def run_system(config_file, vic_exe, test_data_dir, out_dir, driver):
         # ------------------------------------------------------------ #
         # read template global parameter file
-        infile = os.path.join(test_dir, 'samples',
+        infile = os.path.join(test_data_dir + 'image/Stehekin/parameters/', 
```
I don't think we want this change.

yixinmao replied:
If we assume that the global parameter template file is under a certain directory, it should probably be put under `system` subdirectory under `test_dir`, so `infile = os.path.join(test_dir, 'system', test_dict['global_parameter_file'])`

jhamman commented on tests/run_tests.py:
```diff
@@ -339,6 +343,8 @@ def run_system(config_file, vic_exe, test_data_dir, out_dir, driver):
         error_message = ''
 
+        print(vic_exe) 
+
```
remove print statement.

jhamman commented on tests/run_tests.py:
```diff
@@ -351,28 +357,37 @@ def run_system(config_file, vic_exe, test_data_dir, out_dir, driver):
+                if test_dict['check']:
                     # check for nans in the df
                     if 'nonans' in test_dict['check']:
                         check_for_nans(df)
```
this `check_for_nans()` belongs in the classic driver section above.

jhamman commented on tests/run_tests.py:
```diff
@@ -351,28 +357,37 @@ def run_system(config_file, vic_exe, test_data_dir, out_dir, driver):
+            
+                    # check for nans in domain file  
+                    if 'output_nans' in test_dict['check']:
+                        print(test_dict['domain_file'])
```
remove print statement

dgergel replied:
Sorry about these print statements! I forgot to remove them earlier.

jhamman commented on tests/run_tests.py:
```diff
@@ -351,28 +357,37 @@ def run_system(config_file, vic_exe, test_data_dir, out_dir, driver):
+                        print(test_dict['domain_file'])
+                        ds_domain = xr.open_dataset(os.path.join(test_data_dir + 'image/Stehekin/parameters/',
+                                                    test_dict['domain_file']))
+                        ds_output = xr.open_dataset(os.path.join(dirs['results'], 'Stehekin.history.nc')) 
```
We'll want to remove these hard coded paths and get everything from the test config file.

jhamman commented on tests/run_tests.py:
```diff
@@ -767,6 +782,8 @@ def replace_global_values(gp, replace):
     for line in iter(gp.splitlines()):
         line_list = line.split()
+        if line_list == []:
+            continue
```
is this just skipping empty lines? If so, it would be cleaner to write it as:

```Python
if not line_list:
    continue
```

it may be better to go without the continue statement and just do:

```Python
if line_list:
    ... # complete the rest of the for loop
```

yixinmao replied:
Haha sorry I think actually I originally added these two lines... and yes they are for skipping empty lines (otherwise the function breaks at empty lines). Good suggestion.

jhamman commented on tests/system/system_tests.cfg:
```diff
@@ -1,395 +1,451 @@
+#check = None
+#[[options]]
+#BASEFLOW=NIJSSEN2001
+#'''
```
I think we should just remove all these tests until you or @yixinmao can fix the classic driver test system.

dgergel replied:
@jhamman even if we may want to salvage some of them?

jhamman replied:
Yes. That is why we have version control.  We try not to commit commented out code.

jhamman commented on tests/system/system_tests.cfg:
```diff
@@ -1,395 +1,451 @@
+
+[System-check-domainfile-and-nans] 
+driver = image
+test_description = check that NaNs in domain file and output file match 
```
check that the coordinates and mask in the domain and output files match.

yixinmao commented on tests/run_tests.py:
```diff
@@ -351,28 +357,37 @@ def run_system(config_file, vic_exe, test_data_dir, out_dir, driver):
-
-                    # check that each dataframe includes all timestamps
+            # check output files (different tests depending on driver)
+            if driver == "classic":
```
Would it be better to reorder the if statements, say, put `if test_dict['check']` as the highest level, and then check for each possible checks, and then the driver options if needed? Something like this:
```
if test_dict['check']:
        if 'complete' in test_dict['check']:
                do something (maybe another if statement for classic or image driver, if needed)
        if 'nonan' in test_dict['check']:
                do something
```

To me it would be cleaner, and another reason I'm suggesting this is because for some more complicated tests (e.g., the restart tests), maybe only the reading data part is different for different drivers, but the other checking process would be the same. In those cases, it makes more sense to put the distinguishing of drivers in the most inner layer, so that the same checking function can be called.

dgergel replied:
I like this idea. I've standardized the test name to 'output_nans' for both drivers.

yixinmao commented on tests/system/system_tests.cfg:
```diff
@@ -1,395 +1,451 @@
+global_parameter_file = global_param.image.STEHE.feb.txt 
+domain_file = domain.stehekin.20151028.nc
+expected_retval = 0
+check = output_nans
```
To me the check names `nonans` and `output_nans` are kind of confusing. Maybe change for some more self-explaining names?

dgergel replied:
adding more informative name :-)

jhamman commented on tests/system/system_tests.cfg:
```diff
@@ -1,395 +1,451 @@
+[System-check-domainfile-and-nans] 
+driver = image
+test_description = check that NaNs in domain file and output file match 
+global_parameter_file = global_param.image.STEHE.feb.txt 
```
Are you using the global param file in the `examples` directory? We probably need more path info here too.

yixinmao replied:
See my comments above, one reasonable way is to assume the global param template file is in the `test_dir/system`. Although there may be other more generic way to handle the path.

dgergel replied:
I've gone ahead and used @yixinmao's suggestion to use the system subdir in `test_dir`. And yes, it's the global param file from the examples directory, not the old one.

jhamman commented on tests/system/system_tests.cfg:
```diff
@@ -1,395 +1,451 @@
+driver = image
+test_description = check that NaNs in domain file and output file match 
+global_parameter_file = global_param.image.STEHE.feb.txt 
+domain_file = domain.stehekin.20151028.nc
```
I think we'll need more path info here.

yixinmao replied:
Can the path of the domain file be read from the global parameter file?

dgergel replied:
I've put this into the same system subdir in `test_dir` for now.

jhamman commented on tests/system/system_tests.cfg:
```diff
@@ -1,395 +1,451 @@
+domain_file = domain.stehekin.20151028.nc
+expected_retval = 0
+check = output_nans
+[[options]] 
```
Do we need this options section here?  If not, remove it.

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,64 @@
+# In[28]:
+
+file = '/Users/diana/workdir/VIC_tests_20160331/examples/Example-Image-Stehekin-base-case/results/Stehekin.history.nc'
+domainfile = '/Users/diana/Dropbox/UW/VIC_sample_data/image/Stehekin/parameters/domain.stehekin.20151028.nc'
```
remove hard coded paths.

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,64 @@
+# In[29]:
+
+ds = xr.open_dataset(file)
+ds_domain = xr.open_dataset(domainfile)
```
remove these lines as we don't want to run them every time we import this module.

yixinmao commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,64 @@
+    try: 
+        npt.assert_array_equal(ds.lat, ds_domain.lat) and npt.assert_array_equal(ds.lon, ds_domain.lon)
+    except AssertionError: 
+        print("Domains for domain file and output files are different")
```
Here we actually do want to raise an error?

jhamman replied:
Let's rework this a bit:

```Python
def assert_nan_equal(ds_domain, ds_output, match_vars=['mask', 'lat', 'lon']):
    ...
    for var in match_vars:
        npt.assert_all_close(ds[var], ds_domain[var], equal_nan=True)
        # We want an AssertionError to be raised if they don't match.

    ....
```

yixinmao commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,64 @@
+# In[59]:
+
+# calling function 
+assert_nan_equal(ds_domain, ds)
```
What is this line for?

yixinmao commented on tests/run_tests.py:
```diff
@@ -9,13 +9,17 @@
 from tonic.models.vic.vic import VIC, VICRuntimeError  # , read_vic_ascii
 from tonic.io import read_config, read_configobj
 from tonic.testing import check_completed, check_for_nans, VICTestError
+from test_image_driver import assert_nan_equal 
```
Hmm I think this way of calling functions from another file only works when you run the `run_tests.py` script from the directory where the scripts are. @jhamman do we want to put some functions in a separate file, and then call it in some more general way (how?)? Or do we want to just put all the functions in `run_tests.py`?

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,64 @@
+
+# calling function 
+assert_nan_equal(ds_domain, ds)
+
```
remove these lines.

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,64 @@
+    for da in ds_output.data_vars: 
+    
+        # slice first time step of output DataArray 
+        da_sel = ds_output[da].isel(time=0)
```
we want to check all timesteps.

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,64 @@
+            try:
+                npt.assert_array_equal(np.isnan(da_sel.values), np.isnan(mask.values))
+            except AssertionError:
+                print("NaN values in output file and domain file do not match up in %s" % da)
```
This section will need be generalized further. You can assume that the spatial dimension names in the two datasets are the same (e.g. lat/lon) but you won't know what they are (so get them from the mask in the domain file).

I would use `pd.isnull` on the full arrays then compare 2d slices after summing along the non-spatial dims.

Again, if they don't match, we need to raise the `AssertionError`.

yixinmao commented on tests/system/system_tests.cfg:
```diff
@@ -1,395 +1,451 @@
+#'''
+
+[System-check-domainfile-and-nans] 
+driver = image
```
This is more like an issue for later discussion, not necessarily for this PR I guess - so based on our conversation yesterday, I agree with you that it would be cleaner to put the driver option here for each individual system test. But currently driver as well as VIC executable file are still input arguments to `run_tests.py`. I feel this is a little confusing with the driver specification here, but I guess for unit tests we do want to specify the driver from input command line (right?). So maybe we can talk how we want to handle this eventually...

dgergel commented:
@jhamman and @yixinmao I've addressed your comments. The test function may need a bit more work but I wanted to go ahead so that @yixinmao can add restart tests while I'm in the field tomorrow.

yixinmao commented on tests/system/system_tests.cfg:
```diff
@@ -1,395 +1,7 @@
+[System-check-domainfile-and-nans] 
+driver = image
+test_description = check that coordinates and mask in domain and output files match 
+global_parameter_file = global_param.image.STEHE.feb.txt 
```
Hmm I think we want to commit this global parameter template file as well? (also a minor point - I personally think no need to put "feb" in the file name, since this global file should work both for water balance and energy balance I suppose?)

yixinmao commented:
You didn't really change the VIC code, but it seems that the travis tests have failed. Is it possibly because at the beginning of `run_tests.py` it tries to import the other file you added, but it doesn't work unless it's been running under the script directory?

jhamman commented on tests/run_tests.py:
```diff
@@ -351,28 +355,38 @@ def run_system(config_file, vic_exe, test_data_dir, out_dir, driver):
+                            ds_domain = xr.open_dataset(os.path.join(test_dir, 'system', 
+                                                        test_dict['domain_file']))
+                            ds_output = xr.open_dataset(os.path.join(dirs['results'], 
+                                                        'Stehekin.history.nc')) 
```
Instead of hard-coding this file name, can we do a `glob` wildcard (`*.nc`) inside the results directory and check all files present?

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,44 @@
+            if (len(ds_output[da].dims) > 3):  
+                for layer in range(0, len(ds_output[da].nlayer)):
+                    # npt.assert_allclose(da_sel.isel(nlayer=layer), ds_domain['mask'], equal_nan=True)
+                    npt.assert_array_equal(np.isnan(da_sel.isel(nlayer=layer).values), np.isnan(ds_domain['mask'].values))                    
```
break line at 80 chars (PEP8).

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,44 @@
+                    npt.assert_array_equal(np.isnan(da_sel.isel(nlayer=layer).values), np.isnan(ds_domain['mask'].values))                    
+                    
+            else: 
+                # npt.assert_allclose(da_sel, ds_domain['mask'], equal_nan=True)
```
remove commented out code

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,44 @@
+            # check all layers 
+            if (len(ds_output[da].dims) > 3):  
+                for layer in range(0, len(ds_output[da].nlayer)):
+                    # npt.assert_allclose(da_sel.isel(nlayer=layer), ds_domain['mask'], equal_nan=True)
```
remove commented out code

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,44 @@
+    # get lat/lng dim names from domain file 
+    lng_name, lat_name = ds_domain.coords
+
+    match_vars = [lng_name, lat_name] 
```
`coords` will not always unpack in the order you are asserting it will here. It actually doesn't matter so you can generalize this.

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,44 @@
+    for da in ds_output.data_vars: 
+        
+        # check all timesteps 
+        for ts in range(0, len(ds_output.time)):
```
Nothing outright wrong here but the for loop isn't required. I still think using `isull` makes more since here. Some pseudo code for you:

```Python
# Find grid cells that are never nans (that mask should match the domain mask)
da_notnull = da.notnull().all(dim=('time', ..., other_non_space_dims))
# compare the masks
npt.assert_array_equal(da.values, ds_domain['mask'].values)
```

dgergel commented:
@jhamman this should be ready to merge.

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,39 @@
+    """
+
+    # get lat/lng dim names from domain file 
+    match_vars = ds_domain.coords
```
You actually get more than the `dim` names here, you get the coordinates. You should add this distinction to your comment.

I would also change this line slightly to make sure we only check the coordinates that are in the `mask` variable:

`match_vars = ds_domain['mask'].coords`

dgergel replied:
updated.

jhamman commented on tests/test_image_driver.py:
```diff
@@ -0,0 +1,39 @@
+            # check all timesteps 
+            da_isnull_reduced = ds_output[da].isnull().all(dim=('time')) 
+            npt.assert_array_equal(da_isnull_reduced.values, 
+                                    np.isnan(ds_domain['mask'])) 
```
We shouldn't need to specify the `nlayer` dim option or even check if the variable is 4d.

What you want to do is take the difference in the dims between what the `mask` variable has and what your data variable has.  For example:

```Python
In [1]: ds_domain.mask.dims
Out[1]: ('x', 'y')
In [2]: ds_output[da].dims
Out[2]: ('x', 'y', 'time', 'snow_band', 'nlayer')
```

You can do this by taking the symmetric difference:

```Python
In [3]: a = ('x', 'y')
In [4]: b = ('x', 'y', 'time', 'snow_band', 'nlayer')
In [5]: set(a) ^ set(b)
Out[5]: {'nlayer', 'snow_band', 'time'}
```

Then you want to do you `...all(dim=(all_diff_dims)`.  Basically you are collapsing every data variable down to the shape of the mask.

dgergel replied:
So cool! Updated.

jhamman commented:
**Good news** - Everything's passing and this looks ready to go.
**Bad news** - you aren't actually testing anything with this yet.

Let's do two things here:
- use this new tool to test the example output, I think you'll just need to edit this line: https://github.com/UW-Hydro/VIC/blob/develop/tests/examples/examples.cfg#L29
- since you've cleaned out the system tests, let's also run those on travis: you should change this line in both the image and classic drivers: https://github.com/UW-Hydro/VIC/blob/develop/ci/classic.travis#L61

dgergel commented:
@jhamman I've updated the `run_examples` function in `run_tests.py` to include NaN checks (and restructured it so that it matches the `run_system' function for classic/image driver tests).

jhamman commented on tests/run_tests.py:
```diff
@@ -602,7 +616,7 @@ def run_examples(config_file, vic_exe, test_data_dir, out_dir, driver):
         # ------------------------------------------------------------ #
         # read template global parameter file
-        infile = os.path.join(test_dir, 'examples', test_dict['global_parameter_file'])
+        infile = os.path.join(test_dir, 'examples', test_dict['global_parameter_file'])       
```
remove extra whitespace.

jhamman commented on tests/run_tests.py:
```diff
@@ -292,7 +296,7 @@ def run_system(config_file, vic_exe, test_data_dir, out_dir, driver):
         # ------------------------------------------------------------ #
         # read template global parameter file
-        infile = os.path.join(test_dir, 'samples',
+        infile = os.path.join(test_dir, 'system', 
```
remove extra whitespace
